### ISSUE TITLE
Training & updating of the fast weight w0, w1, w2, and lr_fc

### ISSUE THREAD

**Issue by xbillowy** [a1600012888/LaCT#7]
Hi, thank you for sharing the excellent code!

I have some questions about training the TTT module, especifically the update of `w0`, `w1`, `w2`, `self.lr_fc = nn.Linear(dim, self.lr_dim * 3)`, take the NVS task as an example, it seems that the returned updated `w0`, `w1`, `w2` was never used to update `self.w0`, `self.w1`, `self.w2` and `self.lr_fc`.

https://github.com/a1600012888/LaCT/blob/be761fdb37d5928eea16d700c83399eb9deb99ae/lact_nvs/lact_ttt.py#L235-L246

https://github.com/a1600012888/LaCT/blob/be761fdb37d5928eea16d700c83399eb9deb99ae/lact_nvs/model.py#L255-L260

I'm wondering the training process only update regular part of the weights, and the weights of the TTT layer are only temporally updated in each training or testing iteration, and will be discard after the end of the iteration?

**Comment by a1600012888** [comment]
Hi Chen, 

For your question: "do gradients flow across all chunks during training, especially for long sequences"

The answer is **Yes**

At each State-Update(fast weight) Step, the ttt_operation takes a chunk of keys and values and State(fast-weight) updated from last chunk, and perform fast weight update. 

Thus, in the backward pass, the gradients will flow all-the-way.  It's standard BPTT(Backpropagation through time), if you do the detach, it's TBPTT(Truncated Backpropagation Through Time ).

**Comment by a1600012888** [comment]
For NVS, you mentioned about updating with each image autoregressively instead of updating with all input images once:

The major difference would be:
1. updating with each image autoregressively => implies a block-wise causal mask => earlier images cannot see later images.
2. you need to sort the input images so that ajacent images are consecutive frames in the video, otherwise there would be not continuation and locality bias in the data structure (order-less). By default it's random order for lots of  3D dataset.

If images are consecutive frames in the video, I think updating the fast weight at each image autoregressively makes a lot of sense. It would also has lots of applications like online 3D recons, online scanning etc.

**Comment by a1600012888** [comment]
> Another possibility is at the training time, the whole sequence is loaded, so the for-loop [here in lact-ar-video](https://github.com/a1600012888/LaCT/blob/c89091740f462571e190fdb44af4b130b740893f/lact_ar_video/minVid/models/blocks/ar_lact_swa_repeat.py#L104) or [here in lact-llm](https://github.com/a1600012888/LaCT/blob/c89091740f462571e190fdb44af4b130b740893f/lact_llm/lact_model/ttt_operation.py#L123) does the state update and propagation?


You are 100% correct here. For training, the full input seq come in and loop over all chunks!
